Public/Src/Cache/ContentStore/Vsts/DedupReadOnlyContentSession.cs

```diff
@@ -155,14 +172,55 @@ public class DedupReadOnlyContentSession : ContentSessionBase
                 return new PinResult($"DedupStore client requires {RequiredHashType}. Cannot take HashType '{contentHash.HashType}'.");
             }
 
-            try
+            var pinResult = CheckPinInMemory(contentHash);
+            if (pinResult.Succeeded)
             {
-                var pinResult = CheckPinInMemory(contentHash);
-                if (pinResult.Succeeded)
-                {
-                    return pinResult;
-                }
+                return pinResult;
+            }
+
+            VstsBlobIdentifier blobId = ToVstsBlobIdentifier(contentHash.ToBlobIdentifier());
+            VstsDedupIdentifier dedupId = blobId.ToDedupIdentifier();
+
+            var keepUntilResult = await CheckNodeKeepUntilAsync(context, dedupId);
+            if (!keepUntilResult.Succeeded)
+            {
+                // Returned a service error. Fail fast.
+                return new PinResult(keepUntilResult);
+            }
+            else if (!keepUntilResult.Value.HasValue)
+            {
+                // Content not found.
+                return new PinResult(PinResult.ResultCode.ContentNotFound);
+            }
 
+            var timeLeft = keepUntilResult.Value.Value - DateTime.UtcNow;
+
+            if (timeLeft > _ignorePinThreshold)
+            {
+                Tracer.Debug(context, $"Pin ignored bacause keepUntil has remaining time [{timeLeft}] that is greater than the ignorePinThreshold");
+                _dedupCounters[Counters.PinIgnored].Increment();
+                return PinResult.Success;
+            }
+
+            var task = PinImplAsync(context, contentHash);
+
+            if (timeLeft < _pinInlineThreshold)
+            {
+                Tracer.Debug(context, $"Pin inlined bacause keepUntil has remaining time [{timeLeft}] that is less than the pinInlineThreshold");
+                _dedupCounters[Counters.PinInlined].Increment();
+                return await task;
+            }
+
+            task.FireAndForget(context);

```

Comment:
Is there a way to count this case? Say be subtracting the other counts from an overall counter?

Reply:
Definitely possible. I wanted to avoid tracing in this case since I expect most calls to fall into it and I don't want to overwhelm the log with this trace. I could add another counter if you think it's necessary, but it could be possible to just calculate it fairly simply